File: ios/TerraStation/TerraWalletLib/TerraWallet.m
//
//  TerraWallet.m

//
//  Created by Felix on 2020/04/03.
//

#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(TerraWallet, NSObject)

+ (BOOL)requiresMainQueueSetup
{
  return false;  // only do this if your module initialization relies on calling UIKit!
}

RCT_EXTERN_METHOD(getNewWallet: (RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(getNewWalletFromSeed:(NSString *)mnemonic
                  bip:(int)bip
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(sign:(NSDictionary *)message
                  sequence:(NSString *)sequence
                  account_number:(NSString *)account_number
                  chain_id:(NSString *)chain_id
                  hexPrivateKey:(NSString *)hexPrivateKey
                  hexPublicKey:(NSString *)hexPublicKey
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(isValidAddress:(NSString *)address
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

@end

File: ios/TerraStation/TerraWalletLib/TerraWallet.swift
import Foundation

@objc(TerraWallet)
class TerraWallet: NSObject {
    static func moduleName() -> String! {
        return "TerraWallet";
    }

    static func requiresMainQueueSetup() -> Bool {
        //If your module does not require access to UIKit, then you should respond to + requiresMainQueueSetup with NO.
        return false
    }

    //generate
    @objc func getNewWallet(_ resolver: RCTPromiseResolveBlock, rejecter:RCTPromiseRejectBlock) {
        let wallet = TerraWalletSDK.getNewWallet()
        resolver(wallet as NSDictionary)
    }

    //generate from seed
    @objc public func getNewWalletFromSeed(_ mnemonic:String, bip:Int, resolver: RCTPromiseResolveBlock, rejecter:RCTPromiseRejectBlock) {
        let wallet = TerraWalletSDK.getNewWalletFromSeed(mnemonic, bip: bip)
        resolver(wallet as NSDictionary)
    }

    @objc func sign(_ message:Dictionary<String, Any>,
                    sequence:String,
                    account_number:String,
                    chain_id:String,
                    hexPrivateKey:String,
                    hexPublicKey:String,
                    resolver: RCTPromiseResolveBlock,
                    rejecter:RCTPromiseRejectBlock) {

      let body = TerraWalletSDK.sign(message, sequence: sequence, account_number: account_number, chain_id: chain_id, hexPrivateKey: hexPrivateKey, hexPublicKey: hexPublicKey)

      resolver(body as NSDictionary)
    }

    @objc func isValidAddress(_ address:String,
                              resolver: RCTPromiseResolveBlock,
                              rejecter:RCTPromiseRejectBlock) {
        let result = TerraWalletSDK.isValidAddress(address);
        resolver(result);
    }
}
